pkg/ratelimit/limitware/middleware.go

```diff
@@ -0,0 +1,216 @@
+
+// IPKeyFunc returns a function that keys data based on the incoming requests IP
+// address. By default this uses the RemoteAddr, but you can also specify a list
+// of headers which will be checked for an IP address first (e.g.
+// "X-Forwarded-For"). Headers are retrieved using Header.Get(), which means
+// they are case insensitive.
+func IPKeyFunc(headers ...string) KeyFunc {
+	return func(r *http.Request) (string, error) {
+		for _, h := range headers {
+			if v := r.Header.Get(h); v != "" {
+				return v, nil
+			}
+		}
+
+		ip, _, err := net.SplitHostPort(r.RemoteAddr)
+		if err != nil {
+			return "", err
+		}
+		return ip, nil
+	}
+}
+
+// Middleware is a handler/mux that can wrap other middlware to implement HTTP
+// rate limiting. It can rate limit based on an arbitrary KeyFunc, and supports
+// anything that implements limiter.Store.
+type Middleware struct {
+	store      limiter.Store
+	keyFunc    KeyFunc
+	meter      metric.Meter
+	reqCounter metric.Int64Counter
+}
+
+// NewMiddleware creates a new middleware suitable for use as an HTTP handler.
+// This function returns an error if either the Store or KeyFunc are nil.
+func NewMiddleware(s limiter.Store, f KeyFunc) (*Middleware, error) {
+	if s == nil {
+		return nil, fmt.Errorf("store cannot be nil")
+	}
+
+	if f == nil {
+		return nil, fmt.Errorf("key function cannot be nil")
+	}
+
+	meter := global.Meter("github.com/sethvargo/go-limiter")
+	rc, err := meter.NewInt64Counter("request_count", metric.WithDescription("counts number of requests processed by middleware and their status"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Middleware{
+		store:      s,
+		keyFunc:    f,
+		meter:      meter,
+		reqCounter: rc,
+	}, nil
+}
+
+// Handle returns the HTTP handler as a middleware. This handler calls Take() on
+// the store and sets the common rate limiting headers. If the take is
+// successful, the remaining middleware is called. If take is unsuccessful, the
+// middleware chain is halted and the function renders a 429 to the caller with
+// metadata about when it's safe to retry.
+func (m *Middleware) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Call the key function - if this fails, it's an internal server error.
+		key, err := m.keyFunc(r)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		// Take from the store.
+		limit, remaining, reset, ok := m.store.Take(key)
+		resetTime := time.Unix(0, int64(reset)).UTC().Format(time.RFC1123)
+
+		// Set headers (we do this regardless of whether the request is permitted).
+		w.Header().Set(HeaderRateLimitLimit, strconv.FormatUint(limit, 10))
+		w.Header().Set(HeaderRateLimitRemaining, strconv.FormatUint(remaining, 10))
+		w.Header().Set(HeaderRateLimitReset, resetTime)
+
+		// Record request status
+		m.meter.RecordBatch(
```

Comment:
I'm not super familiar with this implementation - does it block? Does it make sense to throw this into a go function so it can run independent and non-blocking of the limiter?

Reply:
I moved to the older implementation, because finding the answer to this was unclear to me. This shouldn't block if I'm reading correctly.